File: station-registry/src/main/java/md/vnastasi/cloud/log/HttpClientLogger.java
package md.vnastasi.cloud.log;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.logging.LoggingHandler;

import java.nio.charset.Charset;

import static io.netty.util.internal.PlatformDependent.allocateUninitializedArray;
import static java.lang.Integer.max;

public class HttpClientLogger extends LoggingHandler {

    public HttpClientLogger(Class<?> clazz) {
        super(clazz);
    }

    @Override
    protected String format(ChannelHandlerContext ctx, String eventName, Object arg) {
        if (arg instanceof ByteBuf) {
            ByteBuf msg = (ByteBuf) arg;
            return decode(msg, msg.readerIndex(), msg.readableBytes(), Charset.defaultCharset());
        }
        return super.format(ctx, eventName, arg);
    }

    private String decode(ByteBuf src, int readerIndex, int len, Charset charset) {
        if (len == 0) return "";

        byte[] array;
        int offset;
        if (src.hasArray()) {
            array = src.array();
            offset = src.arrayOffset() + readerIndex;
        } else {
            array = allocateUninitializedArray(max(len, 1024));
            offset = 0;
            src.getBytes(readerIndex, array, 0, len);
        }
        return new String(array, offset, len, charset);
    }
}

File: disruptions/src/main/kotlin/md/vnastasi/cloud/log/HttpClientLogger.kt
package md.vnastasi.cloud.log

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.logging.LoggingHandler
import io.netty.util.internal.PlatformDependent.allocateUninitializedArray
import java.nio.charset.Charset
import kotlin.math.max
import kotlin.reflect.KClass

class HttpClientLogger(clazz: KClass<*>) : LoggingHandler(clazz.java) {

    override fun format(ctx: ChannelHandlerContext?, eventName: String?, arg: Any?): String {
        if (arg is ByteBuf) {
            return decode(arg, arg.readerIndex(), arg.readableBytes(), Charset.defaultCharset())
        }
        return super.format(ctx, eventName, arg)
    }

    private fun decode(source: ByteBuf, readerIndex: Int, length: Int, charset: Charset): String {
        if (length == 0) return ""

        val array: ByteArray
        val offset: Int

        if (source.hasArray()) {
            array = source.array()
            offset = source.arrayOffset() + readerIndex
        } else {
            array = allocateUninitializedArray(max(length, 1024))
            offset = 0
            source.getBytes(readerIndex, array, 0, length)
        }

        return String(array, offset, length, charset)
    }
}
